test/unit/test_session.py

```diff
@@ -113,6 +117,28 @@ def test_run(self):
             self.assertEqual(session.session_id, job.job_id())
             self.assertEqual(session.backend(), backend)
 
+    def test_run_is_thread_safe(self):
+        """Test the session sends a session starter job once, and only once."""
+        service = MagicMock()
+        api = MagicMock()
+        service._api_client = api
+
+        def _wait_a_bit(*args, **kwargs):
+            # pylint: disable=unused-argument
+            time.sleep(random.random() / 2)
+            return MagicMock()
+
+        service.run = Mock(side_effect=_wait_a_bit)
+
+        session = Session(service=service, backend="ibm_gotham")
+        with ThreadPoolExecutor(max_workers=5) as executor:
+            results = list(map(lambda _: executor.submit(session.run, "", {}), range(5)))
+            wait(results)
+
+        calls = service.run.call_args_list
+        session_starters = list(filter(lambda c: c.kwargs["start_session"] is True, calls))
+        self.assertEqual(len(session_starters), 1)
```

Comment:
Since both threads would encounter the delay in `_wait_a_bit`, it's not a definitive test on whether the lock is working. A better test would be that only the first one has to wait. So in the old path, the second one who doesn't have to wait would have seen the empty job id and proceed to set it before the first is done waiting.

Reply:
That is intended although I see it as a definitive test so let me try to explain what this test is trying to do with the hope you can catch the flaw in my reasoning, and not in my implementation. The test tries to demonstrate that, without the lock, several threads will evaluate the following parameter expression as `True`, starting the session more than once, regardless setting the session id twice.

```python
job = self._service.run(
                ...
                start_session=self._session_id is None,
                ...
            )
```

Making all the threads wait, gives an opportunity to the last ones to reach the parameter evaluation point. Actually, after reading your comments, I realize making them wait for more than [`sys.getswitchinterval()`](https://docs.python.org/3/library/sys.html#sys.getswitchinterval), makes sure more than one will hit the offending line.

I can change the test to wait double of that interval, instead of a random value. I can also reduce the number of threads to 2 which will make the test more deterministic and easy to debug.

Does it make sense?